### DISCUSSION TITLE
How can I configure the portal for a given scenario?

### DISCUSSION THREAD

**Discussion by danilopinotti** [Hieromon/AutoConnect#376] [How can I?] [answered]
Hello!

I'm here after a long time with dozens of trials and errors.
How can I configure the portal to this scenario?
- Portal begins the first time
- When I successfully connect to a network, the portal never comes automatically again.
- The only way to shows the portal again is to clean the configurations manually, like a physical button in my prototype (code at the end of this message).

Currently, I'm using this configurations:
```cpp
this->_portalConfig->autoReset = false;
this->_portalConfig->retainPortal = false;
this->_portalConfig->autoReconnect = true;
this->_portalConfig->reconnectInterval = 1;
this->_portalConfig->immediateStart = false;
this->_portalConfig->beginTimeout = 10000;
```

But I had an unexpected behavior (for me): When the network is down, the ESP shows the portal again.

----
For those who need to reset the configurations manually, follow the code:

```cpp
void DMCNWiFi::resetSettings() {
  Serial.print(F("[WIFI] Resetting Settings..."));
  WiFi.mode(WIFI_STA);
#ifdef ESP32
  WiFi.disconnect(true, true);
#else
  WiFi.persistent(true);
  WiFi.disconnect(true);
  WiFi.persistent(false);
#endif

  AutoConnectCredential credential;
  station_config_t config;
  uint8_t entriesQuantity = credential.entries();

  while (entriesQuantity--) {
    credential.load((int8_t) 0, &config);
    credential.del((const char *) &config.ssid[0]);
  }

  this->_portalConfig->immediateStart = true;
  autoConnect();
  this->_portalConfig->immediateStart = false;
  Serial.println(F("done"));
}
```

**Comment by Hieromon** [comment, marked as answer]
I can't say for sure because I can't see the substance of `autoConnect()` you showed, but maybe you have the steps to register a `_portalConfig` instance with **AutoConnect::config** that should follow the ` this->_portalConfig->immediateStart = false;` setting. It may be leaking. how is it?

```cpp
  this->_portalConfig->immediateStart = true;
  autoConnect(); // -> Does it include AutoConnect::config(_portalConfig)?
  this->_portalConfig->immediateStart = false; 
```

Also, Would you please show me the implementation of the part related to AutoConnect of **DMCNWiFi** class? I will try to configure it to the behavior you expect.

> the portal never comes automatically again.

Similar requirements are raised on [another topic](#375). I didn't expect this, but there seems to be a use case of "enable only the most recently connected access points".
Hmm. Does AutoConnectConfig need an optional configuration for this?

**Reply by danilopinotti** [reply]
@Hieromon thanks for replying.

I need this behavior because, in this project, I will have a dozen nodes fixed in the same environment. When the networks are down (temporarily),  will appear the respective dozen node APs. The ideal scenario for me is: 

1) Node wake up
2) Is the network that the node connected successfully the first time available?
2.1) Yes: Connect -> Send Data -> Disconnect -> Deep sleep for X seconds
2.2) No: Deep sleep for X seconds

If network changed: Reset config manually and then, portal rises and I can connect again on the network.

It's like Sonoff devices works.
___

About my implementation:

I have a wrapper Wifi class named: DMCNWiFi.

In constructor, I set some configs like:
```cpp
DMCNWiFi::DMCNWiFi(char const *apName, char const *apPassword) {
  disableModem(); // Modem is disabled on constructor to power saving

  this->_config = DMCNConfig::get();
  this->apName = apName;
  this->apPassword = apPassword;

  AutoConnectConfig *config = new AutoConnectConfig();
  this->_portalConfig = config;

  this->_portalConfig->apid = apName;
  this->_portalConfig->psk = apPassword;
  this->_portalConfig->autoReset = false;
  this->_portalConfig->retainPortal = false;
  this->_portalConfig->autoReconnect = false;
  this->_portalConfig->reconnectInterval = 1;
  this->_portalConfig->immediateStart = false;
  this->_portalConfig->beginTimeout = 10000;

  this->_portalConfig->title = "DMCN";
  this->_portalConfig->bootUri = AC_ONBOOTURI_HOME;
  this->_portalConfig->homeUri = "/_ac/config";
  this->_portalConfig->menuItems = AC_MENUITEM_CONFIGNEW | AC_MENUITEM_RESET;

  WebServer *webServer;
  AutoConnect *portal = new AutoConnect(*webServer);
  this->_portal = portal;
  this->_portal->config(*this->_portalConfig);
}
```

DMCNWiFi::autoConnect
```cpp
boolean DMCNWiFi::autoConnect() {
  DMCN_DEBUG_INFOLN(F("[WIFI] [BEGIN] Handling Auto Connection"));

  enableModem();
  boolean begin = this->_portal->begin();

  DMCN_DEBUG_INFO(F("[WIFI] Local IP: "));
  DMCN_DEBUG_INFOLN(WiFi.localIP(), false);

  if (WiFi.getMode() & WIFI_AP) {
    WiFi.softAPdisconnect(true);
    WiFi.enableAP(false);
  }

  if (begin) {
    DMCNAux::syncTime();
  }

  DMCN_DEBUG_INFOLN(F("[WIFI] [END] Handling Auto Connection"));
  return begin;
}
```

ResetConfig
```cpp
  this->_portalConfig->immediateStart = true;
  autoConnect(); // -> Does it include AutoConnect::config(_portalConfig)? Yes. I have a single instance of autoConnect setupped on DMCNWiFi constructor.
  this->_portalConfig->immediateStart = false; 
```

The AutoConnect is called in this method:
```cpp
boolean DMCNWiFi::withinConnection(std::function<boolean()> callback) {
  DMCN_DEBUG_INFOLN(F("[WIFI] [BEGIN] Within Connection"));

  boolean isSuccess = (autoConnect() && ensureConnection())
                      ? callback()
                      : false;

  this->_portal->end();
  disableModem();

  DMCN_DEBUG_INFOLN(F("[WIFI] [END] Within Connection"));
  return isSuccess;
}
```

I'm using this approach because the ESP made a lot of things before sending data to the server, and the power consumption is a very serious requeriment.

**Reply by Hieromon** [reply]
Just as I thought, it seems that the reflection of the instance of _portalConfig to AutoConnect is missing.
AutoConnect internally owns a separate AutoConnectConfig instance. The _portalConfig instance acquired by DMCNWiFi is not shared by AutoConnect::config. It is different from the AutoConnectConfig instance referenced by AutoConnect.

```cpp
  this->_portalConfig->immediateStart = true;
  this->_portal->config(*this->_portalConfig);  // It is needed.
  autoConnect();
  this->_portalConfig->immediateStart = false;
  this->_portal->config(*this->_portalConfig);  // It is needed.
```

Or, DMCNWiFi::autoConnect must explicitly copy the _portalConfig instance to AutoConnectConfig inside AutoConnect.

```cpp
boolean DMCNWiFi::autoConnect() {
  DMCN_DEBUG_INFOLN(F("[WIFI] [BEGIN] Handling Auto Connection"));

  enableModem();
  this->_portal->config(*this->_portalConfig);  // It is needed.
  boolean begin = this->_portal->begin();
  ...
}
```

Also, AutoConnect will look up the AutoConnectConfig settings each time the API is called. Its behavior does not force it to be accompanied by AutoConnect::begin. For example, you can use AutoConnect::config to reflect the settings just before calling AutoConnect::handleClient.

**Reply by danilopinotti** [reply]
Thanks! I'll try to implement the #375 solution, then, I notice you of solution.
